runtime/signals.c

```diff
@@ -128,7 +161,19 @@ CAMLexport void caml_leave_blocking_section(void)
   /* Save the value of errno (PR#5982). */
   saved_errno = errno;
   caml_leave_blocking_section_hook ();
-  caml_process_pending_signals();
+
+  /* Some other thread may have switched [caml_signals_might_be_pending]
+     to 0 even though there are still pending signals (masked in the other
+     thread). To handle this case, we force re-examination of all signals.
+
+     Another case where this is necessary (even in a single threaded
+     setting) is when the blocking section unmasks a pending signal:
+     If the signal is pending and masked but has already been
+     examinated by [caml_process_pending_signals], then
+     [caml_signals_might_be_pending] is 0 but the signal needs to be
+     handled at this point. */
+  caml_process_pending_signals(1);
```

Comment:
It is the only place where we pass `force=1`. Can't we just set `caml_signals_might_be_pending` to `1` instead of having a `force` parameter? It is one less parameter to think about and to check whether it is passed correctly or not.

Reply:
As you prefer. This means, however, that we will do the extra pass over all signals  for every call to [caml_process_pending_signals]. If we think performances are important enough for avoiding the system call when possible, then we want to avoid this loop too.

In any case, I don't think trying to avoid either the system call or the scan is worth it in the first place. This function is called in three situations: when the minor heap is full, when entering/leaving a blocking section and when a signal is triggered. The system call is therefore amortized by the heap allocations, the many other mechanisms triggered when entering or leaving blocking sections or the signal handling mechanism.

Reply:
Sorry, I wasn't very clear. What I had in mind was actually this code in`caml_process_pending_signals`:

```c
if (pending) {
  int really_pending = 0;
  for (i = 0; i < NSIG; i++) {
    if (caml_pending_signals[i]) {
      really_pending = 1;
      break;
    }
  }
  pending = really_pending;
}
```

i.e. when `caml_signals_might_be_pending` is set, we check whether it is set appropriately or not. This way we only do extra work when a signal might be pending. I find the code easier to reason about this way.

Reply:
Alright, done. I also squashed the commits.

Reply:
Thanks